### ISSUE TITLE
HTTP/3 Support

### ISSUE THREAD

**Issue by uvd** [hyperium/hyper#1818]
## Current status

HTTP/3 support is on the [roadmap](https://hyper.rs/contrib/roadmap/#http3). There is a working [h3 crate](https://github.com/hyperium/h3), and [reqwest](https://docs.rs/reqwest/latest/reqwest/#unstable-features) has unstable HTTP/3 support.

Much work is still needed, and volunteers are welcome.

Besides the issue tracker (which should store durable decisions and their context), there's this [h3 Discord channel](https://discord.gg/q5mVhMD) to coordinate work.

Here's some of the main steps:

- [ ] Harden the `h3` crate itself, such as fixing any straggling interop issues,
  and filling out the spec conformance tags we use for accountability.
- [ ] Proposal for (initially unstable) `hyper::rt::quic` integration, allowing
  people to bring their own QUIC.
- [ ] Write the `hyper::proto::http3` glue that translates hyper's connection
  patterns with the `h3` crate.

**Comment by shenshouer** [comment]
Any plan for this?

**Comment by quentusrex** [comment]
Since I followed the rabbit hole from Tokio and Tonic to here, the latest work appears to be over here currently:
https://github.com/djc/quinn/pull/734

Any best ways to support the latest efforts on http/3 support?

**Comment by djc** [comment]
I have a branch going on here: https://github.com/djc/hyper/tree/quinn-h3. This adds (completely untested) support for HTTP/3 to the hyper server, but doesn't yet add any API to set it up. I have some ideas on how to do that. The work also more or less depends on #2182 and #2188 (which are currently included in that branch).

**Comment by quentusrex** [comment]
@djc Looks like all of the tickets but https://github.com/hyperium/hyper/pull/2182 were merged. Does this mean that https://github.com/djc/hyper/tree/quinn-h3 is getting close?

**Comment by djc** [comment]
I've been discussing it with @seanmonstar, as far as I can tell he's still figuring out how he wants to go about it.

**Comment by seanmonstar** [comment]
We've put together a proposal here: https://github.com/hyperium/h3/pull/1

**Comment by quentusrex** [comment]
With the proposal merged, what is the best way to see current status towards first MVP of the functionality? and with status, ways to support the effort?

**Comment by djc** [comment]
To the best of my knowledge, work is ongoing in the h3 repo. There is a h3 channel in the Tokio Discord which you might join, as well.

(If by support the effort you mean financially, I might be available to accelerate the efforts on a consulting basis. If that is of interest, shoot me an email.)

**Comment by seanmonstar** [comment]
Good point, I'll see about setting up a milestone with all the issues needed to get the design working. If by support the effort you mean to contribute code, as mentioned, there's an [h3 Discord channel](https://discord.gg/q5mVhMD) to discuss next features, and I'll get a milestone going. There is the general [issue tracker](https://github.com/hyperium/h3/issues), which already has some work split up.

**Comment by quentusrex** [comment]
Congrats on the recent h3 progress(client side request fetching from cloudflare)!

Given the recent activity in h3, and the closing of several of the functionality related issues, what would the hyper next steps look like? I assume there'd be some kind of an h3 release(which would include the minimal public api docs, etc). 

Any easier coding work in view?

**Comment by seanmonstar** [comment]
Yes, an h3 0.1 release would be a necessary step. I've add a few [more issues](https://github.com/hyperium/h3/issues), labels, and a milestone, that should have some "easier" and some more involved work to get there.

Another step to figure out is how to actually provide the support in the `hyper` crate. That's something I haven't thought too much about. I think some of the requirements though, considering the goal of stabilizing hyper into a 1.0 release, would require:

- Not exposing the `h3` library directly, so we can iterate on `h3` without it being breaking changes for `hyper`.
- Making it optional, like `http1` and `http2` are.
- Probably not bundle any particular TLS library with hyper directly.

We can try to figure out both of these things in parallel, so thoughts are welcome!

**Comment by quentusrex** [comment]
Looks like the only issue still attached to the v0.1 milestone is docs related. Is that accurate? https://github.com/hyperium/h3/milestone/1

**Comment by molysgaard** [comment]
IETF has published [RFC 9114](https://www.rfc-editor.org/rfc/rfc9114.html)

**Comment by seanmonstar** [comment]
I've filed an issue in reqwest to explore _using_ the `h3` crate, which would help inform what APIs to expose directly in hyper: https://github.com/seanmonstar/reqwest/issues/1558

**Comment by humbinal** [comment]
Hi, everyone, Is he ready for experimental use now? It is best to use it directly in downstream projects, such as axum, warp...
thanks a lot~

**Comment by mmhdi** [comment]
Hi,
Can you use https://github.com/cloudflare/quiche to support http/3?

**Comment by alikor** [comment]
So any updates on QUIC support?

**Comment by kckeiks** [comment]
Are you accepting PRs for this, perhaps a draft?

**Comment by schizobeyond-tech** [comment]
dont understand how one of most important if not most important features of modern web technologies is still not implemented in most popular Rust server framework especially when `For every second delay in mobile page load, conversions can fall by up to 20%` and HTTP/3 helps with load times and 0RTT a lot..

**Comment by Noah-Kennedy** [comment]
@schizobeyond-tech H3 is still fairly immature as a web technology, and things like congestion control, flow control, IO model, and the standards themselves around QUIC are all evolving still quite rapidly. To compound this, the IO model of H3 is very different than H2 or H1, creating new performance challenges. Not having a dedicated listener socket like with TCP poses a lot of questions and creates a number of challenges. Furthermore, the surface area of HTTP/3 is rapidly evolving in a manner that makes QUIC and H3 support a bit of a moving target (see the introduction of capsules and WebTransport, for example). One of the goals of hyper's H3 support has been support for multiple H3 backends, and this gets much harder because of these challenges.

H3 support in hyper is a work in progress, with much of the work still very much being around answering the question of "how should this be implemented, and what kinds of tradeoffs do we want to make". 

Also, as someone with a lot of experience in this area, H3 performance in the real world is a lot more complicated than you make it out to me. In many cases, even on mobile, H2 can be faster, and we're seeing things like congestion and flow control evolve rapidly to improve the performance of H3 (and even normal TCP as well. Also, zero-RTT doesn't have necessarily the impact you might think all too often. It's very common for real-world performance to be more impacted by bandwidth than the number of RTTs required.

**Comment by QuentinPerez** [comment]
Thank you @Noah-Kennedy for the update. Implementing the client part of HTTP/3 first, could be a viable approach ? Especially considering that many load balancers already support translation between HTTP/3 and HTTP/2.
  
  ```
  ┌────────────┐      ┌──────────┐     ┌──────────┐
│            │      │          │     │          │
│    App     │      │    LB    │     │ Backend  │
│ H3-Client  │─────▶│ H3 -> H2 │────▶│    H2    │
│            │      │          │     │          │
└────────────┘      └──────────┘     └──────────┘
  ```

We have multiple applications running on the AppStore/PlayStore with a networking stack in Rust, having a client-side support for HTTP/3 would be a significant advantage.

**Comment by schizobeyond-tech** [comment]
> @schizobeyond-tech H3 is still fairly immature as a web technology, and things like congestion control, flow control, IO model, and the standards themselves around QUIC are all evolving still quite rapidly. To compound this, the IO model of H3 is very different than H2 or H1, creating new performance challenges. Not having a dedicated listener socket like with TCP poses a lot of questions and creates a number of challenges. Furthermore, the surface area of HTTP/3 is rapidly evolving in a manner that makes QUIC and H3 support a bit of a moving target (see the introduction of capsules and WebTransport, for example). One of the goals of hyper's H3 support has been support for multiple H3 backends, and this gets much harder because of these challenges.
> 
> H3 support in hyper is a work in progress, with much of the work still very much being around answering the question of "how should this be implemented, and what kinds of tradeoffs do we want to make".
> 
> Also, as someone with a lot of experience in this area, H3 performance in the real world is a lot more complicated than you make it out to me. In many cases, even on mobile, H2 can be faster, and we're seeing things like congestion and flow control evolve rapidly to improve the performance of H3 (and even normal TCP as well. Also, zero-RTT doesn't have necessarily the impact you might think all too often. It's very common for real-world performance to be more impacted by bandwidth than the number of RTTs required.

this is just a skill issue on your end, i'll continue using .NET Core instead of hyper. Will check back in 2-5 years

**Comment by seanmonstar** [comment]
@QuentinPerez I've so far suggested building out the client implementation in reqwest, where it can be an internal detail, and flesh how it works, before making some public API in hyper. There is a working (but unstable) feature in reqwest. There's likely a few more things to try out, I could put together an issue over there if interested.

**Comment by NobodyXu** [comment]
> There is a working (but unstable) feature in reqwest.

Can we make it available without having to pass `--cfg reqwest_unstable`?
Since `reqwest` hasn't reached 1.0 yet, it's ok to have new breaking release.

Putting the feature behind `--cfg reqwest_unstable` only make it harder to adopt, I really want to enable http3 support in `cargo-binstall`, however due to the cfg I didn't enable it.

**Comment by QuentinPerez** [comment]
> @QuentinPerez I've so far suggested building out the client implementation in reqwest, where it can be an internal detail, and flesh how it works, before making some public API in hyper. There is a working (but unstable) feature in reqwest. There's likely a few more things to try out, I could put together an issue over there if interested.

I've been using the unstable feature since [its release](https://twitter.com/qperez42/status/1656786662981083140), such as handling media uploads or downloads on GCS. Thus far, it works as expected 🙏  (GCP LB).

Since we use GRPC, I would like to test it with Tonic. If you could provide a straightforward example of an H3 client, integrated into Hyper, it would be awesome. Then I could try integrate it in tonic and test it out with our users and report the issues if any.

**Comment by djc** [comment]
@seanmonstar why not also allow an unstable (guarded by `--cfg`) implementation in hyper?

**Comment by iMonZ** [comment]
Any news here?

**Comment by agavrel** [comment]
Hi,

Any news on this issue?

Love hyper and love quic...

**Comment by dengyunsheng250** [comment]
Hi,

Any news on this issue?

just want to use quic on the axum framework

**Comment by lmtr0** [comment]
Taken from https://docs.rs/hyper/latest/hyper/client/conn/index.html

![Image](https://github.com/user-attachments/assets/ac67f688-8e7b-4552-8986-90f65de42a3e)

Looks like you still need to implement it by hand. quinn is very stable tho

**Comment by dahankzter** [comment]
But surely the protocol has stabilized some more now. Afaict both Nginx and HAProxy supports it now. Is it still the case that it moves too fast to make it worth while to get into Hyper?

**Comment by seanmonstar** [comment]
That's not at all the blocker. Work has been done in the `h3` repo, and it's been tested in reqwest in an unstable feature, (explained in earlier comments). I'll update the top comment with the current status to make it easier to find, but there's no need to leave comments asking if this has progressed, it just pings a bunch of people who are watching.

**Comment by ghost** [comment]
No HTTP/3 yet?

**Comment by eveeifyeve** [comment]
> That's not at all the blocker. Work has been done in the `h3` repo, and it's been tested in reqwest in an unstable feature, (explained in earlier comments). I'll update the top comment with the current status to make it easier to find, but there's no need to leave comments asking if this has progressed, it just pings a bunch of people who are watching.

What about looking into an unstable/experimental feature for hyper? Then people can use h3 in like axum or something as a test then when it becomes more stable introduce it slowly.

Cause I am looking into this more to add support for http3/quic in hyper as experimental.

**Comment by seanmonstar** [comment]
Yea, I did put in the top comment that one of the next steps is to adding unstable support to hyper. I worked on this a little bit a couple weeks ago, and I've now gone ahead and opened a couple WIP pull requests, in case it helps others get started. See #3924 and #3925.

**Comment by XChikuX** [comment]
Not to be THAT guy. But, I've been looking forward to these changes.

Do we have any progress on this?

https://github.com/emmett-framework/granian/issues/550
